src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/timeline/BookingTimelineService.kt

```diff
@@ -209,6 +224,153 @@ class BookingTimelineService(
       .groupBy { it.date }
   }
 
+  private fun applyTrancheAdjustmentLogic(
+    calculation: CalculationResult,
+    adjustments: Adjustments,
+    trancheAndCommencement: Pair<SDSEarlyReleaseTranche, LocalDate?>,
+  ): CalculationResult {
+    if (trancheAndCommencement.first !in listOf(
+        SDSEarlyReleaseTranche.TRANCHE_1,
+        SDSEarlyReleaseTranche.TRANCHE_2,
+      )
+    ) {
+      log.info("No adjustments to apply as tranche is ${calculation.sdsEarlyReleaseTranche}")
+      return calculation
+    }
+
+    return adjustReleaseDatesForTranche(calculation, adjustments, trancheAndCommencement.second!!)
+  }
+
+  private fun adjustReleaseDatesForTranche(
+    calculation: CalculationResult,
+    adjustments: Adjustments,
+    commencementDate: LocalDate,
+  ): CalculationResult {
+    val rulesToHandle = listOf(
+      CalculationRule.SDS_EARLY_RELEASE_ADJUSTED_TO_TRANCHE_1_COMMENCEMENT,
+      CalculationRule.SDS_EARLY_RELEASE_ADJUSTED_TO_TRANCHE_2_COMMENCEMENT,
+      CalculationRule.SDS_STANDARD_RELEASE_APPLIES,
+    )
+
+    val releaseDateTypesToAdjust = listOf(
+      ReleaseDateType.HDCED,
+      ReleaseDateType.ERSED,
+      ReleaseDateType.PED,
+    )
+
+    var updatedCalculation = calculation
+
+    for (rule in rulesToHandle) {
+      for (releaseDateType in releaseDateTypesToAdjust) {
+        updatedCalculation = adjustReleaseDateForType(
+          updatedCalculation,
+          adjustments,
+          commencementDate,
+          releaseDateType,
+          rule,
+        )
+      }
+    }
+
+    return updatedCalculation
+  }
+
+  private fun adjustReleaseDateForType(
+    calculation: CalculationResult,
+    adjustments: Adjustments,
+    commencementDate: LocalDate,
+    releaseDateType: ReleaseDateType,
+    rule: CalculationRule,
+  ): CalculationResult {
+    val applicableRules = calculation.breakdownByReleaseDateType[releaseDateType]?.rules
+
+    if (applicableRules?.contains(rule) == true) {
+      log.info("${releaseDateType.name} before adjustment: ${calculation.dates[releaseDateType]}")
+      val crd = calculation.dates[ReleaseDateType.CRD]!!
+
+      val numberOfDaysToAdjust = calculateAdjustmentDaysAfterCommencement(adjustments, commencementDate, crd)
+
+      if (numberOfDaysToAdjust != 0L) {
```

Comment:
Would it make sense to do this check within the function above, before your start looping over every rule and release date type? The majority of time there won't be any adjustments after tranche commencement.

Reply:
yes it would be more efficient - good spot